code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.ROLE.eq(ProjectConstants.OWNER))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
+    }
+    
+    public Boolean isProjectMember (Long projectId, Long policyAccount, Boolean isAdmin,Set<ExternalId> externalIds){
+        if (projectId == null || externalIds == null) {
+            return false;
+        }
+        if ((policyAccount != null && policyAccount.equals(projectId)) || isAdmin) {
+            return true;
+        }
+        Set<String> roles =  new HashSet<>();
+        roles.add(ProjectConstants.OWNER);
+        roles.add(ProjectConstants.MEMBER);
+        Set<ProjectMemberRecord> projectMembers = new HashSet<>();
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.ROLE.in(roles))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
     }
 
-    @Inject
-    public void setObjectManager(ObjectManager objectManager) {
-        this.objectManager = objectManager;
+    @Override
+    public Account updateProject(Account project, String name, String description) {
+        Map<TableField<AccountRecord, String>, String> properties = new HashMap<>();
+        properties.put(ACCOUNT.NAME, name);
+        properties.put(ACCOUNT.DESCRIPTION, description);
+        int updateCount = create()
+                .update(ACCOUNT)
+                .set(properties)
+                .where(ACCOUNT.ID
+                        .eq(project.getId()))
+                .execute();
+        if(1 != updateCount) {
+            throw new RuntimeException("UpdateProject failed.");
+        }
+        return getAccountById(project.getId());
+    }
+
+    public List<? extends ProjectMember> setProjectMembers(Account project, HashSet<Member> members){
+        List<? extends ProjectMember> previousMembers = getProjectMembers(project.getId());
+        HashSet<Member> otherPreviosMembers = new HashSet<>();
+        for (ProjectMember member: previousMembers){
+            otherPreviosMembers.add(new Member(member));
+        }
```

Comment:
I think we should add a lock on the update here as other requests can be modifying the same data set.